### PR TITLE
hoogle: add nixos service

### PR DESCRIPTION
**Pull request by jb55**

### CODE REVIEW

**Comment by jb55** [comment]
My first service so feedback is appreciated :+1:

File: nixos/modules/services/development/hoogle.nix [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+let
+
+  cfg = config.services.hoogle;
+  ghcWithHoogle = pkgs.haskellPackages.ghcWithHoogle;
+
+in {
+
+  options.services.hoogle = {
+    enable = mkOption {
```
Comment (joachifm) [inline nixos/modules/services/development/hoogle.nix]: `mkEnableOption` is a short-hand for this

Comment (jb55) [inline nixos/modules/services/development/hoogle.nix]: thx, updating

**Comment by joachifm** [comment]
Looks good to me, overall. Will add some suggestions for the option descriptions, but do with them what you want.

File: nixos/modules/services/development/hoogle.nix [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      description = ''
+        Port number Hoogle will be listening to.
+      '';
+    };
+
+    packages = mkOption {
+      default = hp: [];
+      example = "hp: with hp; [ text lens ]";
+      description = ''
+        A function that takes a haskell package set and returns a list of
```
Comment (joachifm) [inline nixos/modules/services/development/hoogle.nix]: I think it's better to focus on what the option is supposed to accomplish, i.e., specifying the packages available to the documentation server, and then say that it's a function that takes the haskellPackages attset &c

File: nixos/modules/services/development/hoogle.nix [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        packages from it.
+      '';
+    };
+
+    haskellPackages = mkOption {
```
Comment (joachifm) [inline nixos/modules/services/development/hoogle.nix]: This option ought to have a default; you can do
`default = pkgs.haskellPackages; defaultText = "pkgs.haskellPackages"`

In the example, perhaps show off using a different package set than the default (i.e., for a specific compiler version).

**Comment by jb55** [comment]
Thanks @joachifm!

**Comment by joachifm** [comment]
Final nitpick: the commit subject should be `nixos: add hoogle service` or something

**Comment by joachifm** [comment]
A suggestion for future development is to run the service as an unprivileged user instead of as root. But ofc, being written in Haskell it won't ever have bugs and it will probably only run on localhost anyway. Still, would be nice, I think.

**Comment by fpletz** [comment]
Hmm, running this as root is probably not a good idea. As this service probably doesn't have any state to keep and modify, maybe nobody would be fitting.

**Comment by joachifm** [comment]
I wish the default was to run as nobody with at least `PrivateTmp` & maybe even without network access, the interface really should be designed so that capabilities had to be granted rather than taken away.

**Comment by Profpatsch** [comment]
@joachifm That might be a good proposal. When defining minimal services, the most sane defaults should be applied. Maybe open an issue?
A problem I see is that most services don’t have integration tests, so we’d need to test them manually afterwards …

**Comment by benley** [comment]
Changing the nixos-wide default is pretty out of scope for this particular change, but I agree, that sounds cool.

**Comment by joachifm** [comment]
Merged at 9c0997a0ef62d178d6bc88aeacc2643481edef9e

**Pull request closed without merging**